source/loader/layers/validation/ur_leak_check.hpp

```diff
@@ -0,0 +1,102 @@
+// Copyright (C) 2023 Intel Corporation
+// SPDX-License-Identifier: MIT
+#ifndef UR_LEAK_CHECK_H
+#define UR_LEAK_CHECK_H 1
+
+#include "backtrace.hpp"
+#include "ur_validation_layer.hpp"
+
+#include <mutex>
+#include <unordered_map>
+#include <utility>
+
+#define MAX_BACKTRACE_FRAMES 64
+
+namespace validation_layer {
+struct RefCountContext {
+  private:
+    struct RefRuntimeInfo {
+        int64_t refCount;
+        std::vector<BacktraceLine> backtrace;
+    };
+
+    std::mutex mutex;
+    std::unordered_map<void *, struct RefRuntimeInfo> counts;
+
+    inline struct RefRuntimeInfo createRefInfo(int refCount) {
+        std::vector<BacktraceLine> backtrace = getCurrentBacktrace();
+
+        return RefRuntimeInfo{refCount, backtrace};
+    }
+
+    std::pair<void *, int64_t> incrementRefCount(void *ptr) {
+        auto it = counts.find(ptr);
+        if (it != counts.end()) {
+            it->second.refCount++;
+            return std::make_pair(it->first, it->second.refCount);
+        } else {
+            counts.emplace(ptr, createRefInfo(1));
```

Comment:
I think we should differentiate between create/retain/release. It will allow you to find issues when retain is called on an object that wasn't created properly.

Reply:
The origin of those calls should be visible in backtrace.

Reply:
Right, but the way you have it, doing:
```
  retain(0xCAFE);
  retain(0xCAFE);
  release(0xCAFE);
  release(0xCAFE);
```
Would be a perfectly valid sequence of calls. That's why I think special-casing `create` here makes sense.

Reply:
I've modified the leak-checking mechanism to differentiate between *Create/*Retain/*Release calls.